whitepapers/0013_ccq.md

```diff
@@ -0,0 +1,350 @@
+If any of the responses fails or times out, the query module will retry periodically for up to one minute. If after a minute some of the per-chain queries were not successful, the query
+module will drop the request.
+
+### Publication of Responses
+
+If and only if the request is successfully processed, the guardian will publish a query response message over P2P using the `ccq_resp` topic. As noted
+previously, the other guardians will not see this message. Only the REST server and possibly third party integrators will see it.
+
+The query response contains both the initial query request and the results. The presence of the request allows the integrator to verify the response is what they are expecting.
+
+The response should be signed with the prefix `query_response_0000000000000000000|`. Note that it is not necessary to have different response prefixes for each environment because
+the responses are signed with the guardian key, which is different between the environments.
+
+### Guardian Configuration
+
+The guardian configuration for CCQ will consist of the following config parameters.
+
+- `ccqEnabled` - if set to `true` then the CCQ feature is enabled. Default is false.
+- `ccqAllowedRequesters` - comma separated list of signer public keys who are allowed to submit query requests. No default.
+- `ccqP2pPort` - local port used to bind the CCQ P2P channel, default is `8996`.
+- `ccqP2pBootstrap` - bootstrap peers for the CCQ P2P channel. No default (but auto generated in tilt).
+- `ccqAllowedPeers` - comma separated list of P2P peer IDs that are allowed to submit query requests.
+
+### No Query Persistence in the Guardian
+
+In order to reduce the storage burden on the guardian node, full responses are not persisted in the guardian. However, to facilitate de-duplication and authorization, some cross-chain query information may be committed to Wormchain.
+
+## CCQ REST Server
+
+Some integrators might choose to communicate directly with the guardians by encoding their own query requests and publishing them to the gossip network. To do so would require their P2P peer ID to be configured on the guardians. Additionally they would need to have a P2P library, which limits their choices (For instance, there does not currently exist a typescript library for P2P). Additionally, the integrator would need to collect and consolidate the responses.
+
+For these and other reasons, a separate REST server is being provided to accept off-chain query requests, submit them to the gossip network, correlate the responses, and return them to the caller.
+
+The CCQ server will be configured to publish and listen on the CCQ P2P network. It's P2P peer ID will be registered with the guardians. The server will join both `ccq_req` and `ccq_resp` but only subscribe to `ccq_resp`.
+
+The server will be configured with a list of supported users. Each user will have an API Key, as well as a list of what queries they are allowed to perform.
```

Comment:
API key allowlist mechanism may change based on conversations this Friday 10/13 (can update whitepaper if so)